include/boost/geometry/algorithms/detail/is_simple/linear.hpp

```diff
@@ -54,128 +57,210 @@ namespace detail { namespace is_simple
+        BOOST_ASSERT(boost::size(m_linestring) > 1);
+        return m_is_closed
+            && turn.method == overlay::method_none
+            && check_segment_indices(turn, boost::size(m_linestring) - 2)
+            && geometry::equals(range::front(m_linestring), turn.point);
     }
-};
-
 
+private:
+    Linestring const& m_linestring;
+    bool const m_is_closed;
+};
 
 template <typename MultiLinestring>
-class is_simple_multilinestring
+class is_acceptable_turn<MultiLinestring, multi_linestring_tag>
 {
 private:
-    class is_acceptable_turn
+    typedef typename boost::range_value<MultiLinestring>::type linestring_type;
+    typedef is_acceptable_turn<linestring_type> base_type;
+
+    template <typename Point, typename Linestring>
+    static inline bool is_boundary_point_of(Point const& point,
+                                            Linestring const& linestring)
     {
-    private:
-        template <typename Point, typename Linestring>
-        static inline bool is_boundary_point_of(Point const& point,
-                                                Linestring const& linestring)
-        {
-            BOOST_ASSERT( boost::size(linestring) > 1 );
-            return
-                !geometry::equals(range::front(linestring),
-                                  range::back(linestring))
-                &&
-                ( geometry::equals(point, range::front(linestring))
-                  || geometry::equals(point, range::back(linestring)) );
-        }
+        BOOST_ASSERT(boost::size(linestring) > 1);
+        return
+            ! geometry::equals(range::front(linestring),
+                               range::back(linestring))
+            &&
+            (geometry::equals(point, range::front(linestring))
+             || geometry::equals(point, range::back(linestring)));
+    }
 
-        template <typename Point, typename Linestring>
-        static inline bool is_closing_point_of(Point const& point,
-                                               Linestring const& linestring)
-        {
-            BOOST_ASSERT( boost::size(linestring) > 1 );
-            return
-                geometry::equals(range::front(linestring),
-                                 range::back(linestring))
-                &&
-                geometry::equals(range::front(linestring), point)
-                ;
-        }
+    template <typename Turn, typename Linestring>
+    static inline bool is_closing_point_of(Turn const& turn,
+                                           Linestring const& linestring)
+    {
+        BOOST_ASSERT(boost::size(linestring) > 1);
+        return
+            turn.method == overlay::method_none
+            &&
+            check_segment_indices(turn, boost::size(linestring) - 2)
+            &&
+            geometry::equals(range::front(linestring), range::back(linestring))
+            &&
+            geometry::equals(range::front(linestring), turn.point)
+            ;
+    }
```

Comment:
@awulkiew Do you have a specific example where there is a mismatch between points and rations. It would be really helpful to have it and use it as test case.